pkg/finality-grandpa/bridge_state.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"sync"
+)
+
+type waker struct {
+	mtx    sync.RWMutex
+	wakeCh chan any
+}
+
+func newWaker() *waker {
+	return &waker{wakeCh: make(chan any, 1000)}
+}
+
+func (w *waker) wake() {
+	w.mtx.RLock()
+	defer w.mtx.RUnlock()
+	if w.wakeCh == nil {
+		return
+	}
+	go func() {
+		select {
+		case w.wakeCh <- nil:
+		default:
+		}
+	}()
+}
+
+func (w *waker) channel() chan any {
+	return w.wakeCh
+}
+
+func (w *waker) register(waker *waker) {
+	w.mtx.Lock()
+	defer w.mtx.Unlock()
+	w.wakeCh = waker.wakeCh
+}
+
+// round state bridged across rounds.
+type bridged[Hash, Number any] struct {
+	inner RoundState[Hash, Number]
+	// registered map[chan State[Hash, Number]]any
+	waker *waker
+	sync.RWMutex
+}
+
+func (b *bridged[H, N]) update(new RoundState[H, N]) {
+	b.Lock()
+	b.inner = new
+	b.waker.wake()
+	b.Unlock()
+}
+
+func (b *bridged[H, N]) get(waker *waker) RoundState[H, N] {
+	b.RLock()
+	defer b.RUnlock()
+	b.waker.register(waker)
+	return b.inner
+}
+
+// A prior view of a round-state.
+type priorView[Hash, Number any] struct {
```

Comment:
are we going to use this priotView with different types of hashes? and numbers?
From my understaning common.Hash works great everywhere and number is almost always uint32. Do we really need generics here. They tend to make code unreadable.

Reply:
This library is utilized in #4144 and #4084 using the primitives introduced in `feat/grandpa` and `refactor/client-db`.  In practice `runtime.Hash` and `runtime.Number` are used ([link](https://github.com/ChainSafe/gossamer/pull/4144/files#diff-9706d68b145ecec3a4c0a2e055da3afdb4dcfe9ac0f95d556ad203e8f533c079)).  In the newer code we utilize `btree.Map` and `btree.Set` pretty extensively in both the standalone and introduced client and primitives code.  This requires that hashes satisfy `constraints.Ordered` which `common.Hash` doesn't satisfy.  Given that this is a supposed to be a library, if the language allows for generics on the number I don't see it as too much added complexity.  Usually if the type that you're defining isn't strict on the type constraints you end up using `any` to be concise if it has a meaningful name.  In practice though `Number` ends being `constraints.Unsigned` since that's defined in  other places in this package.